activerecord/lib/active_record/store.rb

```diff
@@ -290,9 +290,11 @@ def self.as_indifferent_hash(obj)
           def as_regular_hash(obj)
             case obj
             when ActiveSupport::HashWithIndifferentAccess
-              obj.to_h
+              obj.to_hash
             when Hash
-              obj
+              obj.transform_values { |value| value.is_a?(Hash) || value.is_a?(Array) ? as_regular_hash(value) : value }
+            when Array
+              obj.map { |value| value.is_a?(Hash) || value.is_a?(Array) ? as_regular_hash(value) : value }
```

Comment:
This duplicated block could be extracted, but maybe this is short enough that duplication is preferrable to indirection. Matter of taste.